### PR TITLE
add e2e test for network partition

### PR DESCRIPTION
**Pull request by freehan**
e2e test for #8818 and #17741

### CODE REVIEW

**Comment by freehan** [comment]
Assigning to @bprashanth since he has reviewed related PRs.

File: test/e2e/resize_nodes.go [hunk 1]
```diff
@@ -560,6 +566,93 @@ var _ = Describe("Nodes [Disruptive]", func() {
 					}
 				}
 			})
+
+			// What happens in this test:
+			// 	Network connectivity on a node is cut off for 120 seconds to simulate network partition
+			// Expect to observe:
+			// 1. Node is marked NotReady after timeout by nodecontroller (40seconds)
+			// 2. All pods on node are marked NotReady shortly after #1
+			// 3. Node and pods return to Ready after connectivivty recovers
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: I think there's a test that already does 1, 3 in a slightly less destructive way than dropping all traffic+sleep+undropping, can you just have it do 2 as well (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/resize_nodes.go#L501)?

Comment (freehan) [inline test/e2e/resize_nodes.go]: Yeah. I was planning to do that too. But that test is trying to observe pod eviction. Then I could not verify if the pod change from NotReady to Ready after recovery.

Comment (bprashanth) [inline test/e2e/resize_nodes.go]: That test should wait till all addons+static pods+pods on the node at the start of the test are ready, if it does not that is a bug. 

Regarding your pods (that you create in the test), you're not actually marking them ready right? you're just leaving it to the kubelet, so isn't this like any other kubelet-readiness test i.e is there already a test that checks that pods can successfully go from ready -> not ready -> ready today without the netsplit?

Comment (freehan) [inline test/e2e/resize_nodes.go]: I believe the e2e framework waits for all node/pod to be ready before it starts and checks node readiness after each test. 

Before #17741, only kubelet manipulates pod status. Now nodecontroller can also mark pod notready. Hence kubelet cannot assume its pod status cache is always valid, this is fixed in #18410.  If just to test #18410, we still need to change pod status without telling kubelet and see if it could correct it. 

Basically this test wants to test #17741 and #18410

Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Alright, just be prepared if you get assigned a readiness bug :)

File: test/e2e/resize_nodes.go [hunk 2]
```diff
@@ -560,6 +566,93 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+				}
+				node := nodes.Items[0]
+				podOpts := api.ListOptions{FieldSelector: fields.OneTermEqualSelector(client.PodHost, node.Name)}
+				if err = waitForMatchPodsCondition(c, podOpts, "Running and Ready", podReadyTimeout, podRunningReady); err != nil {
+					Failf("Pods on node %s are not ready and running within %v: %v", node.Name, podReadyTimeout, err)
+				}
+				nodeOpts = api.ListOptions{FieldSelector: fields.OneTermEqualSelector("metadata.name", node.Name)}
+
+				By("Set up watch on node status")
+				w, err := c.Nodes().Watch(nodeOpts)
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: This is an error prone way to watch (for many reasons, most notably the watch can break because of a server side timeout). Please use the informer framework, eg; https://github.com/kubernetes/kubernetes/blob/master/test/e2e/daemon_restart.go#L221

File: test/e2e/resize_nodes.go [hunk 3]
```diff
@@ -560,6 +566,93 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+				if err != nil {
+					Failf("Failed to set up watch: %v", err)
+				}
+
+				defer w.Stop()
+
+				By(fmt.Sprintf("Performing temporary network partition on node %s", node.Name))
+				partitionCmd := "nohup sh -c 'sleep 10 && sudo ifdown eth0 && sleep 120 && sudo ifup eth0' >/dev/null 2>&1 &"
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Can you use the partition strategy employed in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/resize_nodes.go#L297 ? it only cuts off the master and doesn't need the sleep logic.

File: test/e2e/resize_nodes.go [hunk 4]
```diff
@@ -560,6 +566,93 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+				Expect(issueSSHCommand(partitionCmd, testContext.Provider, &node)).NotTo(HaveOccurred())
+
+				By("Expect to observe node and pod status change from Ready to NotReady after network partition")
+				timer := time.After(nodeNotReadyTimeout)
+				notReady := false
+				timeout := false
+				var lastNode *api.Node
+				for !notReady && !timeout {
+					select {
+					case event, _ := <-w.ResultChan():
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: With the informer you'd just have an event handler, something like https://github.com/kubernetes/kubernetes/blob/master/test/e2e/density.go#L362

File: test/e2e/resize_nodes.go [hunk 5]
```diff
@@ -560,6 +566,93 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+							}
+						}
+					case <-timer:
+						timeout = true
+					}
+				}
+				if !ready {
+					Fail("Failed to observe node status change from NotReady to Ready")
+				}
+				if err = waitForMatchPodsCondition(c, podOpts, "Running and Ready", podReadyTimeout, podRunningReady); err != nil {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Can you just remember the names of pods and follow the pattern here: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/reboot.go#L209 or does that make things harder?

**Comment by bprashanth** [comment]
Mostly just the one critique about the informer pattern

File: test/e2e/resize_nodes.go (hunk 5 above)
Comment (freehan) [inline test/e2e/resize_nodes.go]: The problem with this approach is that it requires to specify namespaces while I want to check all pods on the node. I could not find a util function to list all pods on a node, that is why I wrote a generic one.

File: test/e2e/resize_nodes.go [hunk 6]
```diff
@@ -560,6 +582,121 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+					}
+
+					By("Expect to observe node and pod status change from NotReady to Ready after network connectivity recovers")
+					timer := time.After(nodeReadyTimeout)
+					ready := false
+					timeout := false
+					for !ready && !timeout {
+						select {
+						case n := <-newNode:
+							if isNodeConditionSetAsExpected(n, api.NodeReady, true) {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: can you log when it's not expected?

File: test/e2e/resize_nodes.go [hunk 7]
```diff
@@ -560,6 +582,121 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+
+				defer func() {
+					By(fmt.Sprintf("Unblock traffic from node %s to the master", node.Name))
+					unblockNetwork(host, master)
+
+					if CurrentGinkgoTestDescription().Failed {
+						return
+					}
+
+					By("Expect to observe node and pod status change from NotReady to Ready after network connectivity recovers")
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: The section between this By and the next feels like it can be isolated into a function (in one case it's notready->ready and the other its ready->notready). Is that easy?

File: test/e2e/resize_nodes.go [hunk 8]
```diff
@@ -324,53 +350,49 @@ func performTemporaryNetworkFailure(c *client.Client, ns, rcName string, replica
 	default:
 		Failf("This test is not supported for provider %s and should be disabled", testContext.Provider)
 	}
-	iptablesRule := fmt.Sprintf("OUTPUT --destination %s --jump REJECT", master)
+	return master
+}
+
+func getNodeExternalIP(node *api.Node) string {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: can you clarify that this returns ip+port for ssh? either via comment or name change

File: test/e2e/resize_nodes.go [hunk 9]
```diff
@@ -560,6 +582,121 @@ var _ = Describe("Nodes [Disruptive]", func() {
...
+				nodes, err := c.Nodes().List(nodeOpts)
+				Expect(err).NotTo(HaveOccurred())
+				filterNodes(nodes, func(node api.Node) bool {
+					return isNodeConditionSetAsExpected(&node, api.NodeReady, true)
+				})
+				if len(nodes.Items) <= 0 {
+					Failf("No node were found Ready: %d", len(nodes.Items))
+				}
+				node := nodes.Items[0]
+				podOpts := api.ListOptions{FieldSelector: fields.OneTermEqualSelector(client.PodHost, node.Name)}
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: you're not actually creating pods right (or did i miss it)? so can you actually confirm this is a non zero number of pods

Comment (freehan) [inline test/e2e/resize_nodes.go]: I did not create pods. I will check to confirm the node has pods on it.

File: test/e2e/resize_nodes.go (hunk 8 above)
Comment (freehan) [inline test/e2e/resize_nodes.go]: Done

File: test/e2e/resize_nodes.go [hunk 10]
```diff
@@ -388,6 +411,27 @@ func performTemporaryNetworkFailure(c *client.Client, ns, rcName string, replica
 	// network traffic is unblocked in a deferred function
 }
 
+func expectNodeReadinss(isReady bool, newNode chan *api.Node) {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Readinss -> Readiness, or did you mean "in seconds" or something?

**Comment by bprashanth** [comment]
LGTM, suggest running this in a loop overnight to make sure it doesn't flake. Once you're confident, feel free to apply the label (after you've addressed the spelling nit).

**Comment by bprashanth** [comment]
LGTM @freehan did you try the overnight stress run?

File: test/e2e/resize_nodes.go [hunk 11]
```diff
@@ -287,27 +295,44 @@ func verifyPods(c *client.Client, ns, name string, wantName bool, replicas int)
...
-// environments.
-func performTemporaryNetworkFailure(c *client.Client, ns, rcName string, replicas int, podNameToDisappear string, node *api.Node) {
-	Logf("Getting external IP address for %s", node.Name)
-	host := ""
-	for _, a := range node.Status.Addresses {
-		if a.Type == api.NodeExternalIP {
-			host = a.Address + ":22"
-			break
-		}
+func blockNetwork(from string, to string) {
```
Comment (bprashanth) [inline test/e2e/resize_nodes.go]: Can you move the block and unblock functions into e2e/utils?
```
type netsplitHelper struct { timeout time.Time }

func (n *netsplitHelper) Block() error {}
func (n *netsplitHelper)Unblock() error {}
```

I'm alright with a follow up pr that does this, I gather many tests can use this functionality (I have one in mind already).

**Comment by freehan** [comment]
@bprashanth 
I did a stress test and found a race condition. There is a slight chance where stopCh and newNode channel were closed but informer still executes onUpdate and causing panic because it is writing to a closed channel.  

I think I can just leave the newNode channel unclosed, since there is GC anyway, plus this is e2e test, not core component. Or do you have any suggestion?

File: test/e2e/resize_nodes.go (hunk 11 above)
Comment (freehan) [inline test/e2e/resize_nodes.go]: I can do a follow up PR. No need to make this one too complex.

**Comment by bprashanth** [comment]
Leaving channels open is fine. It's only dangerous if you have a goroutine listening on the other end for a close signal, because that's a memory leak that'll probably show up on jenkins-master during soak test. If you can confirm that we don't run into that case it's fine to just let the channel get GC'd with a comment.

I mean a goroutine like this will leak its stack frame if you don't close ch:
```go
go func() {
   for {
      val, done := <-ch
      if done {
        return
      }
  }
}
```

**Comment by freehan** [comment]
Alright, run stress test for one more night.

File: test/e2e/resize_nodes.go [hunk 12]
```diff
@@ -287,35 +295,55 @@ func verifyPods(c *client.Client, ns, name string, wantName bool, replicas int)
...
-		// internal IP address instead (i.e. NOT the one in testContext.Host).
-		master = testContext.CloudConfig.MasterName
+		eps, err := c.Endpoints(api.NamespaceDefault).Get("kubernetes")
+		if err != nil {
+			Failf("Fail to get kubernetes endpoinds: %v", err)
+		}
+		if len(eps.Subsets) != 1 || len(eps.Subsets[0].Addresses) != 1 {
+			Failf("There are more than 1 endpoints for kubernetes service: %+v", eps)
+		}
+		master = eps.Subsets[0].Addresses[0].IP
```
Comment (freehan) [inline test/e2e/resize_nodes.go]: Use the IP instead of host name... This is to avoid external DNS dependancy for iptables.

**Comment by freehan** [comment]
Filed issue for GCE DNS problem:  https://buganizer.corp.google.com/u/0/issues/26825573

On our end, use ip instead of hostname.

Will run it for one more night. Fingers crossed...

**Comment by bprashanth** [comment]
LGTM
